lib/pos_blockchain.go

```diff
@@ -18,45 +18,108 @@ import (
 //  4. try to apply the incoming block as the tip (performing reorgs as necessary). If it can't be applied, we exit here.
 //  5. Run the commit rule - If applicable, flushes the incoming block's grandparent to the DB
 func (bc *Blockchain) processBlockPoS(desoBlock *MsgDeSoBlock, currentView uint64, verifySignatures bool) (_success bool, _isOrphan bool, _missingBlockHashes []*BlockHash, _err error) {
-	// 1. Determine if we're missing the parent block of this block. If it's parent exists in the blockIndex,
-	// it is safe to assume we have all ancestors of this block in the block index.
-	// If the parent block is missing, process the orphan, but don't add to the block index.
+	// Start by pulling out the block hash.
+	blockHash, err := desoBlock.Header.Hash()
+	if err != nil {
+		return false, false, nil, errors.Wrap(err, "processBlockPoS: Problem hashing block: ")
+	}
+
+	// Get all the blocks between the current block and the committed tip. If the block
+	// is an orphan, then we store it without validating it. If the block connects to a
+	// committed block other than the committed tip, then we throw it away.
 	lineageFromCommittedTip, err := bc.getLineageFromCommittedTip(desoBlock)
-	if err != nil && err != RuleErrorMissingAncestorBlock {
+	if err == RuleErrorDoesNotExtendCommittedTip {
+		// In this case, the block extends a committed block that is NOT the tip
+		// block. There is no point in storing this block because we will never
+		// reorg to it.
 		return false, false, nil, err
-	}
 
+	}
 	if err == RuleErrorMissingAncestorBlock {
+		// In this case, the block is an orphan that does not connect to any blocks
+		// on our best chain. In this case we'll store it with the hope that we
+		// will eventually get a parent that connects to our best chain.
 		missingBlockHashes := []*BlockHash{desoBlock.Header.PrevBlockHash}
-		blockHash, err := desoBlock.Header.Hash()
-		// If we fail to get the block hash, this block isn't valid at all, so we
-		// don't need to worry about adding it to the orphan list or block index.
+		// FIXME: I sketeched some of these steps but not all...
```

Comment:
Not sure if you added a check for this already. For TxnConnectStatusByIndex we should add the following block validation step:

```
HashBitset(MsgDeSoBlock.TxnConnectStatusByIndex).IsEqual( 
MsgDeSoBlock.Header.TxnConnectStatusByIndexHash)
```

Reply:
I'll add this in. We didn't have it in the MsgDeSoBlock when I was writing all the validations. I'll add this in another PR in the stack.

Reply:
@AeonSw4n - this is addressed in #774 . Please let me know if you have other feedback on this PR.